### ISSUE TITLE
BUG: validation doesn't run in model.save() if model.validate() called first, even if validation fails.

### ISSUE THREAD

**Issue by ryangriggs** [Automattic/mongoose#8468]
If the model.validate() function is called, then the model.save() function is called, the save() function apparently does not validate the data, and proceeds to write the data to the database, even if it would not pass validation.

Full description with example code:
https://stackoverflow.com/questions/59547793/mongoose-saves-invalid-data-without-throwing-validation-errors-if-model-validate

This issue seems to be mitigated by calling the validateSync() function, but even so, it should NEVER write data failing to pass validation to the database.

**Comment by AbdelrahmanHafez** [comment]
I can confirm this, the script below logs: 
```
Validation failed.
Saved successfully.
```

```js
const mongoose = require('mongoose');
const { Schema } = mongoose;

mongoose.connect('mongodb://127.0.0.1:27017/test', { useNewUrlParser: true, useUnifiedTopology: true });

const partSchema = new Schema({
  name: {
    type: String,
    required: true,
    validate: {
      validator: (value) => value !== 'asdf' // Don't allow name to be 'asdf'
    }
  },
  number: {
    type: String,
    required: true,
    validate: {
      validator: (value) => !value.includes(' ') // Don't allow spaces in part number.
    }
  }
});

const productSchema = new Schema({
  name: String,
  parts: [partSchema]
});

const Part = mongoose.model('Part', partSchema);
const Product = mongoose.model('Product', productSchema);

const product = new Product({ name: 'Baseball Bat', parts: [new Part({ name: 'First part', number: '003344' }), new Part({ name: 'Second part', number: '554422' })] });

product.parts.push(new Part({ name: 'No number, so invalid' })); // this one is invalid because no part number is specified (required)
product.parts.push(new Part({ name: 'asdf', number: 'zzzzzaaaa' }));
product.parts.push(new Part({ name: 'bbbb', number: 'with a space' })); // This one is invalid because number has spaces.


product.validate().then(() => console.log('Validation passed.')).catch(() => console.log('Validation failed.'));
product.save().then(() => console.log('Saved successfully.')).catch(() => console.log('Saving failed.'));
```

However, changing the validation/saving bit to the following fixes the issue:

```js
product.validate()
  .then(() => console.log('Validation passed.'))
  .catch(() => console.log('Validation failed.'))
  .then(() => product.save())
  .then(() => console.log('Saved successfully.'))
  .catch(() => console.log('Saving failed.'));
```

For some reason, trying to save with a running async validation the save function ignores the current running async validation.

A couple of ideas to temporarily fix your issue:
1- Just using `.save()` validates the document internally, no need to validate, then save (which validates in itself).
2- Wait for the validation to finish before trying to save (the code snippet fix above).
3- Try async/await, it drastically improves async code flow, and readability.

Meanwhile, I'll be looking into this as soon as I can.

**Comment by vkarpov15** [comment]
The issue here is that you're running async `validate()` twice in parallel: once explicitly when you call `validate()`, and once implicitly when you call `save()`. The internals of how Mongoose validate() works assumes that only one `validate()` runs at any given time because of [`Document#invalidate()`](https://mongoosejs.com/docs/api/document.html#document_Document-invalidate).

The `invalidate()` API works like this: you can call `invalidate()` at any time, and Mongoose will report that error in the next `validate()` call. But that's scoped to a document, not to a call of `validate()`. So running `validate()` twice in parallel inevitably leads to race conditions unless we scrap support for `Document#invalidate()`.

The solution is that we'll throw a `ParallelValidateError` when you call `validate()` twice in parallel, similar to `ParallelSaveError`.

**Comment by tukusejssirs** [comment]
In my case, schema is not validated on `save()` (invalid documents were inserted into the database collection). Also I noticed that Mongoose does not insert validator to MongoDB collection and thus I cannot validate documents directly via MongoDB (e.g. in Mongo Shell)—is there a way to insert the validator there?

That said, I could make the validation work using @AbdelrahmanHafez’s [comment](https://github.com/Automattic/mongoose/issues/8468#issuecomment-570678549), i.e. validating the schema manually. Bellow is a code snippet that I use.


```js
doc.validate()
  .then(() => {
    console.log('[INFO] Validation passed.')

    doc.save()
      .then(() => console.log('[INFO] Document was saved.'))
      .catch(() => console.log('[ERROR] Document was not saved.'))
  })
  .catch(() => console.log('[ERROR] Validation failed.'))
```

**Comment by vkarpov15** [comment]
@tukusejssirs please open a new issue and follow the issue template. Mongoose does run `validate()` before `save()` by default, so your code sample is unnecessary.
